x-pack/plugins/apm/typings/es_schemas/raw/fields/process.ts

```diff
@@ -5,8 +5,8 @@
  */
 
 export interface Process {
-  args: string[];
+  args?: string[];
   pid: number;
-  ppid: number;
-  title: string;
+  ppid?: number;
+  title?: string;
```

Comment:
Good catch. Would be good if we had a mechanism of verifying whether a field is required or not. 

We could for instance make a script that runs the following query on all required fields:
```
POST apm-*-transaction-*/_search?terminate_after=1
{
  "size": 0,
  "query": {
    "bool": {
      "must": [
        {
          "exists": {
            "field": "process"
          }
        }
      ],
      "must_not": [
        {
          "exists": {
            "field": "process.pid"
          }
        }
      ]
    }
  }
}
```

It it returns `count > 0` it means the field is optional because there were found docs without the particular field.

Reply:
This seems like a lot of work for a false sense of security. You can run that script and then I can upload a document to the index that breaks our type's assumption. I'm leaning a lot more toward wanting everything to be an optional `T | T[] | null`, since that's what Elasticsearch is giving us.

Reply:
>  I'm leaning a lot more toward wanting everything to be an optional

Think is, many of the fields are not optional so I think it would be doing us a disservice to treat them as such. The ideal solution would be to talk with the APM Server folks and ask them to publish a spec that we can use to auto-generate the types. But that is not something I expect to have available in the near future.

Reply:
Until now I haven't seen many (if any) bugs due to incorrect types. What problem are we trying to solve?

Reply:
> and then I can upload a document to the index that breaks our type's assumption

Is this something our users are doing? I'd consider that sabotage (to quote @dgieselaar :p)

Reply:
It's something that's a capability of our datastore. If our users do it (they probably don't, but types are a modeling of the possible, not the probable), and the mapping allows it, it's not sabotage.

I don't think this is the most productive forum for this discussion. I think it's something we should think about more deeply, but in this PR I'd like to paste in sample data and have it displayed in a Storybook.

Reply:
@smith I don't think we should strive for correctness with ES mappings. Consider this example:

```json
{
  "properties": {
    "http": {
      "type": "object",
      "properties": {
        "request": {
          "type": "object",
          "properties": {
            "method": {
              "type": "string"
            }
          }
        }
      }
    }
  }
}
```

The type to match this would be:
```ts
type Method = string | string[] | null;
type Request = { method?: Method } | Array<{ method?: Method }>;
type Http = { request?: Request } | Array<{ request?: Request }>;
```

Doesn't seem like that's worth it to me.

Reply:
Yeah, we can have this discussion elsewhere but I think Dario demonstrates how flexible our types will have to be if we don't allow ourself to make any assumptions about the data, other than the raw capabilities of ES.